File: cses/05_range_queries/1143_hotel-queries/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    m = int(data[1])

    # size = lũy thừa 2 nhỏ nhất >= n
    size = 1
    while size < n:
        size <<= 1

    # Max segment tree: mỗi node lưu số phòng trống lớn nhất của đoạn nó quản lí
    tree = [0] * (2 * size)

    # nạp giá trị phòng trống vào các lá
    base = 2
    for i in range(n):
        tree[size + i] = int(data[base + i])

    # dựng tree từ dưới lên
    for i in range(size - 1, 0, -1):
        l = tree[2 * i]
        rr = tree[2 * i + 1]
        tree[i] = l if l >= rr else rr

    out = []
    idx = base + n
    tree_local = tree  # bind cục bộ cho nhanh

    for j in range(m):
        r = int(data[idx + j])
        # max toàn mảng (gốc) < r thì không khách sạn nào đủ phòng -> in 0
        if tree_local[1] < r:
            out.append(b'0')
            continue
        # đi xuống (descent) tìm lá trái nhất có giá trị >= r, ưu tiên con trái
        pos = 1
        while pos < size:
            left = 2 * pos
            if tree_local[left] >= r:
                pos = left
            else:
                pos = left + 1
        # pos là chỉ số lá; khách sạn (1-indexed)
        out.append(str(pos - size + 1).encode())
        # point update: trừ r rồi cập nhật tổ tiên
        tree_local[pos] -= r
        pos >>= 1
        while pos >= 1:
            a = tree_local[2 * pos]
            b = tree_local[2 * pos + 1]
            tree_local[pos] = a if a >= b else b
            pos >>= 1

    sys.stdout.buffer.write(b' '.join(out) + b'\n')


main()

File: cses/05_range_queries/1143_hotel-queries/solution.cpp
#include <algorithm>
#include <cstddef>
#include <iostream>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::size_t n = 0;
    std::size_t m = 0;
    std::cin >> n >> m;

    // size = lũy thừa 2 nhỏ nhất >= n (kích thước tầng lá của segment tree)
    std::size_t size = 1;
    while (size < n) {
        size *= 2;
    }

    // Max segment tree: nạp số phòng trống vào các lá rồi dựng cây từ dưới lên
    std::vector<long long> tree(2 * size, 0);
    for (std::size_t i = 0; i < n; ++i) {
        std::cin >> tree[size + i];
    }
    for (std::size_t node = size; node-- > 1;) {
        tree[node] = std::max(tree[2 * node], tree[2 * node + 1]);
    }

    for (std::size_t query = 0; query < m; ++query) {
        long long rooms = 0;
        std::cin >> rooms;

        std::size_t answer = 0;
        // Nếu max toàn mảng (gốc) < rooms thì không khách sạn nào đủ -> in 0
        if (tree[1] >= rooms) {
            // Đi xuống (descent) tìm lá trái nhất có số phòng trống >= rooms:
            // ưu tiên con trái nếu con trái còn đủ phòng
            std::size_t node = 1;
            while (node < size) {
                if (tree[2 * node] >= rooms) {
                    node *= 2;
                } else {
                    node = 2 * node + 1;
                }
            }

            // node là chỉ số lá; khách sạn đánh số từ 1
            answer = node - size + 1;

            // Point update: trừ rooms vào lá rồi cập nhật lại max các tổ tiên
            tree[node] -= rooms;
            for (node /= 2; node >= 1; node /= 2) {
                tree[node] = std::max(tree[2 * node], tree[2 * node + 1]);
            }
        }

        if (query > 0) {
            std::cout << ' ';
        }
        std::cout << answer;
    }
    std::cout << '\n';
    return 0;
}
